idaes/core/initialization/general_hierarchical.py

```diff
@@ -0,0 +1,156 @@
+#################################################################################
+# The Institute for the Design of Advanced Energy Systems Integrated Platform
+# Framework (IDAES IP) was produced under the DOE Institute for the
+# Design of Advanced Energy Systems (IDAES), and is copyright (c) 2018-2021
+# by the software owners: The Regents of the University of California, through
+# Lawrence Berkeley National Laboratory,  National Technology & Engineering
+# Solutions of Sandia, LLC, Carnegie Mellon University, West Virginia University
+# Research Corporation, et al.  All rights reserved.
+#
+# Please see the files COPYRIGHT.md and LICENSE.md for full copyright and
+# license information.
+#################################################################################
+"""
+Initializer class for implementing Hierarchical initialization routines for
+IDAES models with standard forms (e.g. units with 1 control volume)
+"""
+from pyomo.environ import Block
+from pyomo.common.config import ConfigValue
+
+from idaes.core.initialization.initializer_base import InitializerBase
+from idaes.core.solvers import get_solver
+from idaes.core.util import to_json, from_json, StoreSpec
+import idaes.logger as idaeslog
+
+__author__ = "Andrew Lee"
+
+
+class SingleControlVolumeUnitInitializer(InitializerBase):
+    """
+    This is a general purpose Initializer object for unit models which
+    have a single Control Volume named 'control_volume'.
+    """
+
+    CONFIG = InitializerBase.CONFIG()
+    CONFIG.declare(
+        "solver",
+        ConfigValue(
+            default=None,
+            description="Solver to use for initialization",
+        ),
+    )
+    CONFIG.declare(
+        "solver_options",
+        ConfigValue(
+            default={},
+            description="Dict of options to pass to solver",
+        ),
+    )
+
+    def initialization_routine(self, model: Block):
+        """
+        Common initialization routine for models with standard form.
+
+        Args:
+            model: Pyomo Block to be initialized
+
+        Returns:
+            Pyomo solver results object
+        """
+        if not hasattr(model, "control_volume"):
+            raise TypeError(
+                f"Model {model.name} does not appear to be a standard form unit model. "
+                f"Please use an Initializer specific to the model being initialized."
+            )
+
+        # Get logger
+        _log = self.get_logger(model)
+
+        sub_initializers = {}
+        for sm in model._initialization_order:
```

Comment:
Could you suggest how you would see this working? I agree that the initialization order is more a property of initialization than the model, but for this to be recorded on the initializer it means the initializer must be tied to a specific model instance and registered with that model so that add-ons can be registered as they are created (so that we know what is htere). This means it is a lot harder to experiment with different initializers (you need to rebuild the model) and means you cannot reuse initializers. Plus, I am not sure that it makes much difference in that sense - the model can use initiazation_order to record what is there, but the initializer can choose what to do with it,

1. All `ProcessBlock` types would have an initialization order (which would default to `[self]`). The intent is that all types of `ProcessBlocks` should behave the same way.
2&3. The private attribute is a hold over from the previous implementation, and can be made public. The intent is that users can manipulate this if required.

Reply:
Do you have an example of a model (possibly with add-ons) that would require a non-trivial initialization order? And am I correct to assume that the only Initializers that will use this initialization order are hierarchical ones? If so I could try to make a more specific proposal. Based on the requirements of
1. initialization objects not tied to specific models,
2. the ability to register add-ons as created,
3. and the ability to experiment with different initializers,
my first thought would be to attempt to dynamically compute an initialization order with some simple rule like "unit models, then costing blocks, then flowsheet"

Reply:
@Robbybp Not yet - the most we have at this point are models with costing, and we haven;t really done up a proper test case for them yet either. You are correct however that (so far at least) we only envision `initialization_order` being used in hierarchical type models. Note that we also specifically want to support the capability for users to inject custom code into the initialization order.

In response to your specific comments:

1. In general yes, I think it would be better to have Initialization objects ties to specific model instances (specific model types on the other hand are fine).
2. I think we do need to register the add-ons as they are created, as this would be our only reliable way of knowing what has been added.
3. I am not sure how general that sort of rule would be - for one there could be cases where you want to initialize and add-on before the main unit (e.g. something that computes heat and mass transfer coefficients). I also think we can exclude flowsheets for this (for now at least) - they add even more complexity in that they can be fairly tightly coupled (e.g. costing has unit level blocks and a flowsheet level block).

Reply:
My first thought is that something that computes property coefficients should find a way to be implemented as a property package (would require being added to the model at construction time) or a transformation that is responsible for its own initialization. But it's hard to plan for something like this without an example.

Reply:
Heat and mass transfer coefficients are not thermophysical properties however, as they also depend on model geometry.

To expand a bit on my envisioned (still hypothetical) case: we could have a model (e.g. packed bed) which contains`Vars` for the heat and mass transfer coefficients, which are then used in the material balances. However, there could be a number of different correlations for calculating these (e.g. different packing types), and these correlations could be implemented as an "add-on". I.e. a sub-block which contains some constraints to calculate the heat and mass transfer coefficients on the main model (along with any parameters and variables required by the correlations). In this case, we might want to initialize these correlations with some initial guesses to get initial values for the coefficients which we then use to initialize the column model.

Reply:
In the case of transfer coefficients, these could be implemented as a unit model. But the motivation for having them as an "add-on" is to allow experimenting with different of these add-ons without declaring new classes? And the additional variables/constraints/expressions of this add-on would live on their own block? Attached to the flowsheet or the unit model they are applied to?

Reply:
They could be implemented as part of the unit model, and that is how we currently do it (via derived classes), but my thought was this would make it easier to implement and experiment with different correlations (especially for initialization), as the correlation developer would not need to worry about how this got integrated into the main model. I would think the best place for these to live would be attached to the unit model they are applied to. However, this is all just ideas for now.

Reply:
My first thought for something like this is that it could hopefully be covered by some semi-automatic initialization strategy (i.e. initialized by a block triangular subroutine within a hierarchical strategy). However, this may be easier said than done.

Reply:
That is also my hope, especially as many of these would be fairly simple, but I wanted to make sure we had the framework in place to allow for more complex initialization if required.